Adding rocJPEG to rocPyDecode

essamROCm opened the pull request:
The PR is ready for review.
Documentation will start by the DOCs team; 2 required JIRA tickets were issued.

essamROCm committed:
Adding rocJPEG to rocPyDecode

essamROCm committed:
Preparing source files

essamROCm committed:
Adding PY paramaters and API interface

essamROCm committed:
Adding enum structures types to PyBind11 API

essamROCm committed:
Adding all the rocJPEG APIs with PY sample startup code

essamROCm committed:
Adding Destroy APIs and sample call

essamROCm committed:
Adding rocJPEG APIs & calling from the sample

essamROCm committed:
Merge branch 'develop' into ea/Adding_rocPyJPEG

essamROCm committed:
Adding more rocJPEG APIs and their pybinding plus sample code

essamROCm committed:
Adding HIP Alloc-Free-Check APIs

essamROCm committed:
Adding more rocJPEG APIs

AryanSalmanpour commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,77 @@
+        return self.jpegdec.rocPyAllocHipDeviceMemory(num_channels)
+
+    def rocPyJpegDecode(self, rocjpeg_handle, rocjpeg_stream_handle):
+        return self.jpegdec.rocPyJpegDecode(rocjpeg_handle, rocjpeg_stream_handle)
```
Please also add the `rocPyJpegDecodeBatched` function in this file.

essamROCm replied:
I will add it next commit, needed for the 2nd sample (..batched.py).

essamROCm replied:
Done.

AryanSalmanpour reviewed (commented):
Please add a sample for JPEG batch decoding (e.g., joegdecodebatched.py) in the samples/rocjpeg folder.

essamROCm commented:
> Please add a sample for JPEG batch decoding (e.g., joegdecodebatched.py) in the samples/rocjpeg folder.

Yes, planned, right after I finish the 1st jpegdecoder.py which is almost done.

essamROCm committed:
Py Sample complete, with one sig fault in hip free

essamROCm committed:
Correcting widths and heights to be an array

essamROCm committed:
Moving all storage-variables to user space out of the main class

essamROCm committed:
Splitting Utils class out and adding hip python Alloc-Free calls

essamROCm committed:
Clean up variables names, extra empty lines

essamROCm committed:
Two methods to call HIP

essamROCm committed:
Return all changed values to the user to re-use

AryanSalmanpour commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,76 @@
+        num_components, subsampling, widths, heights = self.jpegdec.rocPyJpegGetImageInfo(rocjpeg_handle, rocjpeg_stream_handle)
+        return num_components, subsampling, widths, heights
+
+    def rocPyInitDecodeParams(self, decode_params, output_format, crop_rect):
```
What is the purpose of this function? The `decode_params` (i.e., `RocJpegDecodeParams` struct) contains `output_format` and `crop_rectangle`, which serve as inputs for the `rocJpegDecode` and `rocJpegDecodeBatched` APIs. The user populates this struct and submits it to the API.

essamROCm replied:
The inner structure don't have a TAG, that is a problem for python binding in order to use the structure inner members I must redefine it adding the inner structures TAGs.

AryanSalmanpour commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,76 @@
+    def rocPyJpegDecode(self, decode_params, rocjpeg_handle, rocjpeg_stream_handle, output_image):
+        return self.jpegdec.rocPyJpegDecode(decode_params, rocjpeg_handle, rocjpeg_stream_handle, output_image)
+
+    def PyInitHipDevice(self, device_id):
```
Please move the `PyInitHipDevice` function from its current location to the `util` class. Additionally, remove both `rocPyFreeHipDeviceMemory` functions from the decoder class. Memory allocation and deallocation should be handled directly in the sample files. The decoder class should only include the `rocPYJpeg*` functions.

essamROCm replied:
Done, moved in utils.

AryanSalmanpour commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,313 @@
+
+    # init decode params
+    is_roi_valid = False
+    roi_width, roi_height, decode_params = jpegdecode.rocPyInitDecodeParams(decode_params, output_format, crop_rect)
```
Please pass the decode_params directly to the `rocPyJpegDecode` API. There is no need to call `rocPyInitDecodeParams` here. The `rocPyInitDecodeParams` should be removed from the decoder class.

essamROCm replied:
Done, removed the rocPyInitDecodeParams

essamROCm committed:
Moving PyInitDecodeParams from decode to Utils class

essamROCm committed:
Making sure address are passed correctly nor issues, removing un-needed code

essamROCm committed:
Clean up sample, change ptr type attempting to fix sig fault

essamROCm committed:
wrapping void * in a safe capsules, destroy APIs implied in the capsules, preparing batch API

essamROCm committed:
adding debug code, making sure alloc ret an address

essamROCm committed:
testing with original sample code

essamROCm committed:
Adding Batched Jpeg Decode Sample and related APIs

essamROCm committed:
Fix for build error

essamROCm committed:
Adding missing variables to the sample removing debug code

essamROCm committed:
Rewrote the batch sample with roi modified its APIs

essamROCm committed:
Merge branch 'develop' into ea/Adding_rocPyJPEG

essamROCm committed:
Jpeg class is a standalone class, Utils class not inheriting from Jpeg class

essamROCm committed:
Merge branch 'develop' into ea/Adding_rocPyJPEG

essamROCm committed:
Removing un-necessary code and include headers

essamROCm committed:
Merge branch 'develop' into ea/Adding_rocPyJPEG

essamROCm committed:
Fix for correct type of parameter and init var variable in py sample

AryanSalmanpour commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,309 @@
+        for i in range(num_channels):
+            if output_image.channel[i] is not None:
+                rocjpeg_status = jpegutils.PyFreeHipDeviceMemory(output_image.channel[i])
+        print(f"Successfully freed allocated MEM, channels count: {num_channels}\n")
```
Please remove lines 205 to 208. This code was for debugging and is no longer needed. Memory allocation and freeing are properly handled in lines 167 to 169 above.

essamROCm replied:
Removed now.

AryanSalmanpour commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,309 @@
+        if(num_channels <= 0):
+            print(f"ERROR: Failed to get the channel pitch and sizes {num_channels}")
+            exit
+        print(f"channel_sizes: {channel_sizes}")
```
Please delete line 161, as it was included only for debugging purposes.

essamROCm replied:
Removed in last commit.

essamROCm committed:
Removing extra debug code from sample

essamROCm committed:
Remove debug code from batched sample, NO SIG FAULT in both batched and single image decode

AryanSalmanpour commented on CMakeLists.txt:
```diff
@@ -159,6 +160,7 @@ set(CMAKE_SKIP_BUILD_RPATH true)
 # Find dependencies
 find_package(HIP QUIET)
 find_package(rocDecode QUIET)
+find_package(rocJPEG QUIET)
```
After line [85](https://github.com/ROCm/rocPyDecode/blob/develop/CMakeLists.txt#L85) in the current CMakeLists.txt, add the following line: `list(APPEND CMAKE_PREFIX_PATH ${ROCM_PATH}/lib/cmake)`. Then, change `find_package(rocJPEG QUIET)` to `find_package(ROCJPEG QUIET)`.

essamROCm replied:
These changes are now committed.

essamROCm committed:
Correcting batch process logic, adding validation to params in samples

AryanSalmanpour commented on CMakeLists.txt:
```diff
@@ -183,32 +185,40 @@ endif()
+endif()
+# rocJPEG
+if(NOT ROCJPEG_FOUND)
+    set(ROCJPEG_FOUND 0)
```
Please do not set `ROCJPEG_FOUND` to 0 or 1. This variable is automatically set to 1 if ROCJPEG is found; otherwise, it will be 0. Please rename `ROCJPEG_FOUND` to `ENABLE_ROPYCJPEG`.

essamROCm replied:
Will be in next commit.

AryanSalmanpour commented on CMakeLists.txt:
```diff
@@ -183,32 +185,40 @@ endif()
+    message("-- ${Yellow}NOTE: ${PROJECT_NAME} library optional rocJPEG, Not Found ${ColourReset}")
+else()
+    set(ROCJPEG_FOUND 1)
+    message("-- ${Yellow}NOTE: ${PROJECT_NAME} library optional rocJPEG, Was Found ${ColourReset}")
```
Please remove lines 200 - 201 AND replace with the code below: We only enable the rocPyJpeg if the rocJPEG version is >= 0.11.1 by setting the `ENABLE_ROCPYJPEG` to 1; otherwise, we don't.

   ```
if(${ROCJPEG_VERSION_MAJOR} VERSION_GREATER_EQUAL "0" AND ${ROCJPEG_VERSION_MINOR} VERSION_GREATER_EQUAL "11" AND ${ROCJPEG_VERSION_PATCH} VERSION_GREATER_EQUAL "1")
        set(ENABLE_ROCPYJPEG 1)
        message("-- Found rocJPEG version ${ROCJPEG_VERSION_MAJOR}.${ROCJPEG_VERSION_MINOR}.${ROCJPEG_VERSION_PATCH} ${ColourReset}")
    else()
        set(ENABLE_ROCPYJPEG 0)
        message("-- ${Yellow}${PROJECT_NAME} -- requires rocJPEG version >= 0.11.1${ColourReset}")
    endif()
```

essamROCm replied:
Will be in next commit.

AryanSalmanpour commented on CMakeLists.txt:
```diff
@@ -223,12 +233,22 @@ if(BUILD_ROCPYDECODE)
+    file(GLOB sources src/*.cpp  ${ROCM_PATH}/share/rocdecode/utils/*.cpp ${ROCM_PATH}/share/rocdecode/utils/rocvideodecode/*.cpp ${ROCM_PATH}/share/rocdecode/utils/ffmpegvideodecode/*.cpp)
+
+    # rocJPEG
+    if(ROCJPEG_FOUND)
```
Please change this line to `if(ROCJPEG_FOUND AND ENABLE_ROCPYJPEG)`

essamROCm replied:
Done.

AryanSalmanpour commented on CMakeLists.txt:
```diff
@@ -223,12 +233,22 @@ if(BUILD_ROCPYDECODE)
+    if(ROCJPEG_FOUND)
+        include_directories(${ROCJPEG_INCLUDE_DIR}
+            ${ROCM_PATH}/share/rocjpeg/samples
+            ${ROCM_PATH}/include/rocjpeg)
```
Please remove `${ROCM_PATH}/include/rocjpeg` from `include_directories` because the `${ROCJPEG_INCLUDE_DIR}` has the correct path.

essamROCm replied:
Done.

AryanSalmanpour commented on CMakeLists.txt:
```diff
@@ -314,22 +334,40 @@ if(BUILD_ROCPYDECODE)
+                message(STATUS "Building and installing rocPyDecode for ${PYTHON_VERSION}")
+                install(FILES ${SO_DST_1} DESTINATION ${CMAKE_INSTALL_LIBDIR} COMPONENT runtime)
+
+                if(ROCJPEG_FOUND)
```
Please change this line to `if(ROCJPEG_FOUND AND ENABLE_ROCPYJPEG)`

essamROCm replied:
Done

AryanSalmanpour commented on cmake/FindrocJPEG.cmake:
```diff
@@ -0,0 +1,89 @@
+################################################################################
```
Please remove `FindrocJPEG.cmake` from the cmake folder. This file is not required for finding the rocjpeg package.

essamROCm replied:
Done.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg.h:
```diff
@@ -0,0 +1,69 @@
+#define PY_ROC_JPEG_PYBIND11_HEADER
+#pragma once
+
+#include "rocjpeg.h"
```
please change this line to ` #include "rocjpeg/rocjpeg.h"`

essamROCm replied:
Done.

essamROCm committed:
Modifying code for review comments, CMakeLists.txt, remove .cmake file

AryanSalmanpour commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,244 @@
+            file_data, file_size = read_image(file_path)
+
+            # Parse stream
+            jpegdecode.rocPyJpegStreamParse(file_data, file_size, stream_handles[idx])
```
Please check the return status of the `rocPyJpegStreamParse` API. If it is not `ROCJPEG_STATUS_SUCCESS`, handle the situation appropriately. Please follow the same logic as in the C++ jpegDecodeBatched sample [here](https://github.com/ROCm/rocJPEG/blob/develop/samples/jpegDecodeBatched/jpegdecodebatched.cpp#L122).

essamROCm replied:
Done, the same logic as in the C++ jpegDecodeBatched is followed in last commit.

AryanSalmanpour commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,244 @@
+            jpegdecode.rocPyJpegStreamParse(file_data, file_size, stream_handles[idx])
+
+            # Get image info
+            _, subsampling, widths, heights = jpegdecode.rocPyJpegGetImageInfo(decode_handle, stream_handles[idx])
```
Please check the returned `subsampling` and if it is `ROCJPEG_CSS_411` or `ROCJPEG_CSS_UNKNOWN` handle it appropriately. Please follow the same logic as in the C++  jpegDecodeBatched sample [here](https://github.com/ROCm/rocJPEG/blob/develop/samples/jpegDecodeBatched/jpegdecodebatched.cpp#L146)

essamROCm replied:
Done, the same logic as in the C++ jpegDecodeBatched is followed in last commit.

AryanSalmanpour commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,244 @@
+            _, subsampling, widths, heights = jpegdecode.rocPyJpegGetImageInfo(decode_handle, stream_handles[idx])
+
+            # Get channel sizes & allocate memory
+            num_channels, channel_sizes = jpegutils.PyGetChannelPitchAndSizes(decode_params_batch[idx], subsampling, widths, heights, output_images[idx])
```
The `PyGetChannelPitchAndSizes` should return a status and it needs to be checked here. Please see [here](https://github.com/ROCm/rocJPEG/blob/develop/samples/jpegDecodeBatched/jpegdecodebatched.cpp#L161)

essamROCm replied:
Done, the same logic as in the C++ jpegDecodeBatched is followed in last commit.

AryanSalmanpour commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,244 @@
+
+        # Batched Decode
+        start_time = datetime.datetime.now()
+        jpegdecode.rocPyJpegDecodeBatched(decode_handle, stream_handles_batch, len(current_batch_files), decode_params_batch, output_images)
```
Please check the return status of the `rocPyJpegDecodeBatched` API and handle it properly. If the status is not ROCJPEG_STATUS_SUCCESS, print an error message.

essamROCm replied:
Done, the same logic as in the C++ jpegDecodeBatched is followed in last commit.

AryanSalmanpour commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,244 @@
+                print(f"Saved: {output_file_name}")
+
+    print("\nBatched JPEG decoding completed.\n")
+
```
At the end of the loop, after all of the images in a folder have been processed, the application should report the following totals: the total number of decoded images, the total number of skipped images, the total number of images that could not be parsed, the total number of images with YUV 4:1:1 chroma subsampling, and the total number of images with unknown chroma subsampling. As well as the Average processing time per image (ms), Average decoded images per sec (Images/Sec), and Average decoded images size (Mpixels/Sec). Please follow the same logic as in the C++ jpegDecodeBatched sample [here](https://github.com/ROCm/rocJPEG/blob/develop/samples/jpegDecodeBatched/jpegdecodebatched.cpp#L224)

sample output from C++ sample:

```
Total decoded images: 56
Total skipped images: 4 ,total images that cannot be parsed: 1 ,total images with YUV 4:1:1 chroam subsampling: 2 ,total images with unknwon chroam subsampling: 1
Average processing time per image (ms): 2.58281
Average decoded images per sec (Images/Sec): 387.175
Average decoded images size (Mpixels/Sec): 101.712
Decoding completed!
```

essamROCm replied:
Done, the same logic as in the C++ jpegDecodeBatched is followed in last commit.

AryanSalmanpour commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,244 @@
+
+    print("\nBatched JPEG decoding completed.\n")
+
+
```
At the end of the program, please call the `rocPyJpegDestroy` and `rocPyJpegStreamDestroy` APIs to release the resources appropriately. Please see [here](https://github.com/ROCm/rocJPEG/blob/develop/samples/jpegDecodeBatched/jpegdecodebatched.cpp#L263). You would also need to do a clean-up for output_image if it is not freed previously. See [here](https://github.com/ROCm/rocJPEG/blob/develop/samples/jpegDecodeBatched/jpegdecodebatched.cpp#L253)

essamROCm replied:
Done, the same logic as in the C++ jpegDecodeBatched is followed in last commit. The Destroy part is implied/moved-to the creator destructor on rocPyJpeg

essamROCm replied:
The APIs are back, and the logic in original code is used now.

AryanSalmanpour commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,321 @@
+                print(f"Average decoded images per sec (Images/Sec): {images_per_sec:.2f}")
+                print(f"Average decoded images size (Mpixels/Sec): {mpixels_per_sec:.2f}")
+
+    # end
```
At the end of the program, please call the `rocPyJpegDestroy `and `rocPyJpegStreamDestroy `APIs to release the resources appropriately. Please see [here](https://github.com/ROCm/rocJPEG/blob/develop/samples/jpegDecode/jpegdecode.cpp#L231). You would also need to do a clean-up for `output_image` if it is not freed previously. See [here](https://github.com/ROCm/rocJPEG/blob/develop/samples/jpegDecode/jpegdecode.cpp#L194)

essamROCm replied:
Done in last commit.

AryanSalmanpour commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,321 @@
+        print("Decoding started, please wait! ... ")
+        start_time = datetime.datetime.now()
+
+        rocjpeg_status = jpegdecode.rocPyJpegDecode(decode_params, decode_handle, stream_handle, output_image) # Call the JPEG decode
```
Please check the return status of the `rocPyJpegDecode `API and handle it properly. If the status is not ROCJPEG_STATUS_SUCCESS, print an error message.

essamROCm replied:
Done, the same logic as in the C++ jpegDecodeBatched is followed in last commit.

AryanSalmanpour commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,321 @@
+            else:
+                exit
+
+        num_channels, channel_sizes = jpegutils.PyGetChannelPitchAndSizes(decode_params, subsampling, widths, heights, output_image)
```
The `PyGetChannelPitchAndSizes` should return a status and it needs to be checked here. Please see [here](https://github.com/ROCm/rocJPEG/blob/develop/samples/jpegDecodeBatched/jpegdecodebatched.cpp#L161)

essamROCm replied:
Done, the same logic as in the C++ jpegDecodeBatched is followed in last commit.

essamROCm committed:
Proper error check examining every status return value in all calls

essamROCm committed:
All status check and error handling moved to sample, batch sample now identical to c++ sample, still sig fault issues to be resolved

essamROCm committed:
Logic fix was causing sig fault, now fixed

essamROCm committed:
batch py sample output is identical to c++ sample output, except in measured time

essamROCm committed:
Proper error handling in single image decoding python sample

AryanSalmanpour reviewed (changes requested):
Hi @essamROCm, Please also make the changes below. Thanks.

1 - in the main CMakeLists.txt in this [line  ](https://github.com/ROCm/rocPyDecode/blob/develop/CMakeLists.txt#L395)`set(ROCPYDECODE_DEBIAN_PACKAGE_LIST  "rocdecode-dev, libdlpack-dev, python3-dev, python3-pip")` add `rojpeg-dev` as a dependency

2- in the main CMakeLists.txt in this [line ](https://github.com/ROCm/rocPyDecode/blob/develop/CMakeLists.txt#L397)` set(ROCPYDECODE_RPM_PACKAGE_LIST  "roc-decode-devel, python3-devel, python3-pip")` please add the `rocjpeg-devel` as a dependency


3- Please add CTests for rocJPEG in the [tests/CMakeLists.txt](https://github.com/ROCm/rocPyDecode/blob/develop/tests/CMakeLists.txt)

Please add a check to execute the rocJPEG CTEST if the version of rocJPEG is 0.11.1 or higher.


```
# 12 - jpegdecode_test
list(APPEND CMAKE_PREFIX_PATH ${ROCM_PATH}/lib/cmake)
find_package(ROCJPEG QUIET)
if(ROCJPEG_FOUND)
  if(${ROCJPEG_VERSION_MAJOR} VERSION_GREATER_EQUAL "0" AND ${ROCJPEG_VERSION_MINOR} VERSION_GREATER_EQUAL "11" AND ${ROCJPEG_VERSION_PATCH} VERSION_GREATER_EQUAL "1")
      add_test(NAME jpeg_decode_python
            COMMAND ${Python3_EXECUTABLE} 
            ${ROCM_PATH}/lib/pyRocVideoDecode/samples/rocjpeg/jpegdecode.py
            -i ${ROCM_PATH}/share/rocjpeg/images/
            WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
        set_property(TEST jpeg_decode_python PROPERTY ENVIRONMENT "PYTHONPATH=${ROCM_PATH}/lib:$PYTHONPATH")

  endif()
endif()
```

essamROCm committed:
Adding rocjpeg as dependency, adding ctest for jpegdecode

essamROCm commented:
rocjpeg added as a dependency, ctest added for jpegdecode.

essamROCm committed:
Adding batched sample in the ctest list

AryanSalmanpour commented on CMakeLists.txt:
```diff
@@ -183,32 +185,45 @@ endif()
+    set(ENABLE_ROCPYJPEG 0)
+    message("-- ${Yellow}NOTE: ${PROJECT_NAME} library optional rocJPEG, Not Found ${ColourReset}")
+else()
+    if(${ROCJPEG_VERSION_MAJOR} VERSION_GREATER_EQUAL "0" AND ${ROCJPEG_VERSION_MINOR} VERSION_GREATER_EQUAL "11" AND ${ROCJPEG_VERSION_PATCH} VERSION_GREATER_EQUAL "1")
```
Please replace this line

`if(${ROCJPEG_VERSION_MAJOR} VERSION_GREATER_EQUAL "0" AND ${ROCJPEG_VERSION_MINOR} VERSION_GREATER_EQUAL "11" AND ${ROCJPEG_VERSION_PATCH} VERSION_GREATER_EQUAL "1")
`

with 

```
    
if(${ROCJPEG_VERSION_MAJOR} VERSION_GREATER_EQUAL "0" OR
      (${ROCJPEG_VERSION_MAJOR} VERSION_EQUAL "0" AND ${ROCJPEG_VERSION_MINOR} VERSION_GREATER_EQUAL "11") OR
      (${ROCJPEG_VERSION_MAJOR} VERSION_EQUAL "0" AND ${ROCJPEG_VERSION_MINOR} VERSION_EQUAL "11" AND ${ROCJPEG_VERSION_PATCH} VERSION_GREATER_EQUAL "1"))
```

```

essamROCm replied:
Done.

AryanSalmanpour commented on CMakeLists.txt:
```diff
@@ -223,12 +238,21 @@ if(BUILD_ROCPYDECODE)
+            ${ROCM_PATH}/share/rocjpeg/samples)
+        set(LINK_LIBRARY_LIST ${LINK_LIBRARY_LIST} ${ROCJPEG_LIBRARY})
+        file(GLOB pyfiles_jpeg pyRocJpegDecode/*.py pyRocJpegDecode/*.pyi)
+        file(GLOB include_jpeg src/rocjpeg/*.h ${ROCM_PATH}/share/rocjpeg/samples/*.h) # include the "rocjpeg_samples_utils.h"
```
please change `${ROCM_PATH}/share/rocjpeg/samples/*.h` with `${ROCM_PATH}/share/rocjpeg/samples/rocjpeg_samples_utils.h`

also, please remove the comment `# include the "rocjpeg_samples_utils.h"`

essamROCm replied:
Done.

AryanSalmanpour commented on tests/CMakeLists.txt:
```diff
@@ -138,6 +138,19 @@ if(Python3_FOUND AND ROCPYDECODE_PYBIND_SCRIPTS)
+  list(APPEND CMAKE_PREFIX_PATH ${ROCM_PATH}/lib/cmake)
+  find_package(ROCJPEG QUIET)
+  if(ROCJPEG_FOUND)
+    if(${ROCJPEG_VERSION_MAJOR} VERSION_GREATER_EQUAL "0" AND ${ROCJPEG_VERSION_MINOR} VERSION_GREATER_EQUAL "11" AND ${ROCJPEG_VERSION_PATCH} VERSION_GREATER_EQUAL "1")
```
Please replace this line

`if(${ROCJPEG_VERSION_MAJOR} VERSION_GREATER_EQUAL "0" AND ${ROCJPEG_VERSION_MINOR} VERSION_GREATER_EQUAL "11" AND ${ROCJPEG_VERSION_PATCH} VERSION_GREATER_EQUAL "1")
`

with 

```
    
if(${ROCJPEG_VERSION_MAJOR} VERSION_GREATER_EQUAL "0" OR
      (${ROCJPEG_VERSION_MAJOR} VERSION_EQUAL "0" AND ${ROCJPEG_VERSION_MINOR} VERSION_GREATER_EQUAL "11") OR
      (${ROCJPEG_VERSION_MAJOR} VERSION_EQUAL "0" AND ${ROCJPEG_VERSION_MINOR} VERSION_EQUAL "11" AND ${ROCJPEG_VERSION_PATCH} VERSION_GREATER_EQUAL "1"))
```

essamROCm replied:
Done.

AryanSalmanpour commented on tests/CMakeLists.txt:
```diff
@@ -138,6 +138,32 @@ if(Python3_FOUND AND ROCPYDECODE_PYBIND_SCRIPTS)
+  endif()
+  # 13 - jpegdecodebatched_test
+  list(APPEND CMAKE_PREFIX_PATH ${ROCM_PATH}/lib/cmake)
+  find_package(ROCJPEG QUIET)
```
Please remove lines 155 and 156 to find rocjpeg package again. Both are already included above in line 142-143

you can simply add the new add_test command right after line 151 above

essamROCm replied:
Done.

essamROCm committed:
Correcting if version in 2 Cmake, removing repeated lines

AryanSalmanpour commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,72 @@
+        subsampling, widths, heights, status = self.jpegdec.rocPyJpegGetImageInfo(rocjpeg_handle, rocjpeg_stream_handle)
+        return subsampling, widths, heights, status
+
+    def rocPyJpegDecode(self, decode_params, rocjpeg_handle, rocjpeg_stream_handle, output_image):
```
Please keep the argument order similar to the C++ API. Move the `decode_params` to be the fifth argument, right before `output_image`.

essamROCm replied:
Done.

AryanSalmanpour commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,72 @@
+        status = self.jpegdec.rocPyJpegDecode(decode_params, rocjpeg_handle, rocjpeg_stream_handle, capsule)
+        return status
+
+    def rocPyJpegDecodeBatched(self, decode_handle_capsule, stream_capsules, batch_size, in_decode_params, destinations_capsule):
```
Please rename `decode_handle_capsule` to `decode_handle`
`stream_capsules` to `stream_handles`
`destinations_capsule` to `destinations`

essamROCm replied:
Done.

AryanSalmanpour commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,72 @@
+        ctypes.pythonapi.PyCapsule_New.restype = ctypes.py_object
+        capsule = ctypes.pythonapi.PyCapsule_New(ctypes.byref(destinations_capsule), b'RocJpegImage', None)
+        status = self.jpegdec.rocPyJpegDecodeBatched(decode_handle_capsule, stream_capsules, batch_size, in_decode_params, capsule)
+        return status
```
Please include the `rocPyJpegDestroy` and `rocPyJpegStreamDestroy` APIs, as they are currently missing from the decoder class. Both APIs must be used at the end of the` jpegdecode.py` and `jpegdecodebatched.py` samples for resource cleanup

essamROCm replied:
Done.

LakshmiKumar23 commented on CMakeLists.txt:
```diff
@@ -314,22 +340,40 @@ if(BUILD_ROCPYDECODE)
                 # Add a custom command to rename the shared library only after it's built
-                add_custom_command(TARGET ${TARGET_NAME_VERSIONED} POST_BUILD COMMAND ${CMAKE_COMMAND} -E copy ${SO_SRC} ${SO_DST} COMMENT "Renaming shared library after build: ${SO_SRC} -> ${SO_DST}" VERBATIM)
-                #install rocPyDecode .so
+                add_custom_command(TARGET ${TARGET_NAME_VERSIONED} POST_BUILD COMMAND ${CMAKE_COMMAND} -E copy ${SO_SRC_1} ${SO_DST_1} COMMENT "Renaming shared library after build: ${SO_SRC_1} -> ${SO_DST_1}" VERBATIM)                
```
what is `SO_SRC_1` , `SO_SRC_2` etc? Is it `.so` for rocDecode and rocJPEG? Can we give it more meaningful names?

essamROCm replied:
Done.

LakshmiKumar23 commented on pyRocJpegDecode/__init__.py:
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2018 - 2023 Advanced Micro Devices, Inc. All rights reserved.
```
please change to `2025 -`

essamROCm replied:
Done.

LakshmiKumar23 commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,72 @@
+# Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
same as above

essamROCm replied:
Done.

LakshmiKumar23 commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,72 @@
+        self.jpegdec = rocpyjpeg.PyRocJpegDecoder()
+
+    def rocPyJpegCreate(self, backend, device_id):
+        handle, status = self.jpegdec.rocPyJpegCreate(jpegt.RocJpegBackend(backend), device_id)
```
please assign `self.backend = backend` & `self.device_id = device_id` and pass these to the fucntion. This is standard Python programming. This way it can be access in class `decoder` wherever needed in other methods/functions as well. Similarly for other functions as well

LakshmiKumar23 replied:
please redo line 46 to `self.jpeg_handle, status = self.jpegdec.rocPyJpegCreate(jpegt.RocJpegBackend(backend), device_id)`. This can be reused across all methods and does not need to be passed to function. Same with `self.rocjpeg_stream_handle` can be a member function. Can use this across all functions and need not be exposed

essamROCm replied:
Done.

LakshmiKumar23 commented on pyRocJpegDecode/types.py:
```diff
@@ -0,0 +1,106 @@
+# Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
`2025`

essamROCm replied:
Done

LakshmiKumar23 commented on pyRocJpegDecode/utils.py:
```diff
@@ -0,0 +1,61 @@
+# Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
`2025`

essamROCm replied:
Done.

LakshmiKumar23 commented on pyRocJpegDecode/utils.py:
```diff
@@ -0,0 +1,61 @@
+        capsule = ctypes.pythonapi.PyCapsule_New(ctypes.byref(output_image), b'RocJpegImage', None)
+        return self.jpegutils.PySaveImage(decode_params, image_save_path, width, height, subsampling, capsule)
+
+    def PyInitHipDevice(self, device_id):
```
we do not need a wrapper class for this. We can use `hip-python` directly, which we use in rocPyDecode here already.
https://github.com/ROCm/rocPyDecode/blob/b5912d48c167abc080660e381754401a6e3b2919/samples/videodecodeperf.py#L9
https://github.com/ROCm/rocPyDecode/blob/b5912d48c167abc080660e381754401a6e3b2919/samples/videodecodeperf.py#L141

LakshmiKumar23 commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,344 @@
+    output_file_path = args.output
+    device_id = args.device
+    rocjpeg_backend = args.backend
+    crop_rect = args.crop_rect
```
please check validity `crop_rect is not None` here to avoid copy to another `crp_rct` in the function

essamROCm replied:
Done.

LakshmiKumar23 commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,344 @@
+    decode_params = jdec.PyRocJpegDecodeParams()
+    decode_params.output_format = get_format(output_format)
+    crp_rct = [0,0,0,0]
+    if(crop_rect is not None):
```
please do this check in `__main__`

essamROCm replied:
Done.

LakshmiKumar23 commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,344 @@
+
+        num_channels, channel_sizes, status = jpegutils.PyGetChannelPitchAndSizes(decode_params, subsampling, widths, heights, output_image)
+        if(status != jpegt.ROCJPEG_STATUS_SUCCESS):
+            print(f"Failure - GetChannelPitchAndSizes Status: {status}")
```
please exit() after failure

essamROCm replied:
Done.

LakshmiKumar23 commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,344 @@
+            prior_channel_sizes[i] = channel_sizes[i]
+
+        if (save_images):
+            image_save_path = output_file_path
```
not needed, since it's already there in line 216

essamROCm replied:
The statement if (is_dir) could evaluate to FALSE, so the image_save_path  wouldn't have assigned a value before used in SaveImage() - which produce runtime error.

LakshmiKumar23 commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,357 @@
+    decode_params = jdec.PyRocJpegDecodeParams()
+    decode_params.output_format = get_format(output_format)
+    crp_rct = [0,0,0,0]
+    if(crop_rect is not None):
```
same comment as other sample

essamROCm replied:
Done.

LakshmiKumar23 commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,344 @@
+import pyRocJpegDecode.decoder as jdec
```
add copyright

essamROCm replied:
Done.

LakshmiKumar23 replied:
The whole block should be added as other files

LakshmiKumar23 commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,357 @@
+import pyRocJpegDecode.decoder as jdec
```
add copyright

essamROCm replied:
Done.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg.h:
```diff
@@ -0,0 +1,69 @@
+/*
+Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
2025

essamROCm replied:
Done.

essamROCm committed:
Arguments redone in original order
renamed variables in API decoder
returned the 'Destroy' APIs back
renamed Cmake vars to have meaning
corrected dates in multiple files
added self assignment in the decoder API
checked crop rect at main

essamROCm commented:
@AryanSalmanpour @LakshmiKumar23 
Hi Aryan, Lakshmi, Changes were made and pushed for your latest requests. The PR is ready to review now. Thank you.

AryanSalmanpour commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,365 @@
+                    print(f"Failure - FreeHipDeviceMemory Status: {status}")
+                    sys.exit(1)
+                else:
+                    output_image.channel[i] = 0
```
Please move lines 234 to 241 outside of the for loop. They should be relocated to after line 268 in the clean-up section below.

essamROCm replied:
It seems the whole block starting line 233 is also in the loop by mistake, should be outside.

essamROCm replied:
Moved out of the loop as well as the final reporting block.

essamROCm committed:
Fix for jpegdecode loop logic, error messages

rrawther commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,80 @@
+import ctypes
+
+
+class PyRocJpegImage(ctypes.Structure):
```
why this class only has 2 members to represent image? What about all other attributes

essamROCm replied:
It is a structure inherited from rocJPEG, pybind11/Python knows class instead of structure, it is by design has 2 members to represent the image.

rrawther replied:
If we have a separate class for Image, it is better to keep all image attributes here which is obtained from get_image_info

essamROCm commented:
@rrawther @LakshmiKumar23 
Hi Rajy, Lakshmi, do you have more questions or request for code change, or should I go ahead and merge this PR?

rrawther commented on pyRocJpegDecode/utils.py:
```diff
@@ -0,0 +1,61 @@
+    def PyInitHipDevice(self, device_id):
+        return self.jpegutils.PyInitHipDevice(device_id)
+
+    def PyAllocHipDeviceMemory(self, channel_size):
```
Is channel_size the size of memory allocated? If so please make it generic name because the function doesn't care what type of memory needs to be allocated

essamROCm replied:
I will change that in next commit.

rrawther commented on pyRocJpegDecode/utils.py:
```diff
@@ -0,0 +1,61 @@
+        output_ptr, status = self.jpegutils.PyAllocHipDeviceMemory(channel_size)
+        return output_ptr, status
+
+    def PyFreeHipDeviceMemory(self, output_image_channel):
```
same comment here

essamROCm replied:
I will change it in next commit.

LakshmiKumar23 commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,80 @@
+        self.stream_handle, status = self.jpegdec.rocPyJpegStreamCreate()
+        return self.stream_handle, status
+
+    def rocPyJpegStreamParse(self, file_data, length, jpeg_stream_handle):
```
The `jpeg_stream_handle` is already  available in the class as you created in line 48 and line 52. Do not need to pass handle and expose this in Python. Can use the one created above appropriately

essamROCm replied:
@AryanSalmanpour, please correct me here. The idea of keeping the handle in the class and re-use it can work for the 'Decoder' which I can re-use as you suggested for it is one handle for the whole session, I will change that in the next commit. However, the 'Stream(s)' handle can be so many, as many as a batch size can be, which is user defined. So, you cannot rely on the last stream handle to reuse. The stream handles must remain in the user hand to pass them in calls.

essamROCm replied:
I can explain more in details if required in our meeting

LakshmiKumar23 commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,80 @@
+        status = self.jpegdec.rocPyJpegStreamParse(file_array, length, jpeg_stream_handle)
+        return status
+
+    def rocPyJpegGetImageInfo(self, rocjpeg_handle, rocjpeg_stream_handle):
```
same as above

essamROCm replied:
Same response as above.

LakshmiKumar23 commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,80 @@
+        subsampling, widths, heights, status = self.jpegdec.rocPyJpegGetImageInfo(rocjpeg_handle, rocjpeg_stream_handle)
+        return subsampling, widths, heights, status
+
+    def rocPyJpegDecode(self, rocjpeg_handle, rocjpeg_stream_handle, decode_params, output_image):
```
Both handles are already created in the class above. Use as member function

essamROCm replied:
I can re-use the decoder handle as it is just one per session. I will change that in next commit.

rrawther commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,366 @@
+        return None, 0
+    try:
+        # Open file in binary mode
+        with open(file_path, "rb") as input_file:
```
Python has better os file functions. We don't need to replicate every line of code from c++. For e.g, os.path.get_size() is one line which will give the file size. Please rewrite this file to be more python friendly

essamROCm replied:
Will change that function. The whole sample can latter be optimized to fewer lines if different PRs.

rrawther commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,366 @@
+    # Create/Init RocJpegDecodeParams
+    decode_params = jdec.PyRocJpegDecodeParams()
+    decode_params.output_format = get_format(output_format)
+    decode_params.crop_rectangle.left = crop_rect[0]
```
same here. please assign all coordinates in single line

essamROCm replied:
Will change that next commit.

rrawther commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,380 @@
+        return None, 0
+    try:
+        # Open file in binary mode
+        with open(file_path, "rb") as input_file:
```
suggest rewriting this with fewer lines of code. We don't need to replicate every line of c++ lib

essamROCm replied:
Will change this function next commit.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg.h:
```diff
@@ -0,0 +1,69 @@
+#include <pybind11/functional.h>
+#include <pybind11/stl.h>
+#include <pybind11/numpy.h>
+// #include <pybind11/embed.h>
```
remove unused lines

essamROCm replied:
Will be removed next commit.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpegdecode.cpp:
```diff
@@ -0,0 +1,252 @@
+/*
+Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
2025

essamROCm replied:
Will change in next commit

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg.cpp:
```diff
@@ -0,0 +1,153 @@
+/*
+Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
2025

essamROCm replied:
Will change in next commit

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpegdecode.h:
```diff
@@ -0,0 +1,65 @@
+/*
+Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
2025

essamROCm replied:
Will change in next commit

LakshmiKumar23 commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,366 @@
+            if not os.path.isdir(output_file_path): # out must be dir
+                print("ERROR: for passing folder path as input, you must pass an existing folder path as output.")
+                sys.exit(1)
+    if(device_id < 0):
```
you can restirct this in the `parser` function itself with `choices` as keyword. Something like this `choices=range(0,8)`. Similarly for others. All these lines can be removed from both samples

essamROCm replied:
Done in next commit.

rrawther reviewed (changes requested):
@essamROCm I would like to go over the changes in a meeting.

essamROCm committed:
Changes made for the requests in PR review conversation

essamROCm commented:
> @essamROCm I would like to go over the changes in a meeting.

Hi @rrawther Rajy
I have made all the requested changes in this commit: https://github.com/ROCm/rocPyDecode/pull/170/commits/67eb28114be6e59af6885506c3b3fe206f940df9 & https://github.com/ROCm/rocPyDecode/pull/170/commits/775fcdf419fd5bb96fcf3a0bbb8911c401d4f1e5

Please suggest suitable time for you and the team to go over the changes in a meeting.

essamROCm committed:
CMake fix, and adding full block of copyrights

kiritigowda committed:
Merge branch 'develop' into ea/Adding_rocPyJPEG
